### PR TITLE
[orchestration] Add support for versioned SLC documents

### LINKED ISSUE #414: Add epic grouping to CSD
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Adapt LevaDocgen to provide in CSD document the epic information to add an extra level for grouping requirements. This will be done as the official CSD template allows, adding subsections inside the existing sections. 

**Describe alternatives you've considered**
N/A

**Additional context**
Requirement from one of the projects. 

See also https://github.com/opendevstack/ods-document-generation-templates/issues/15 

FYI @metmajer  @jorge-romero  @hrcornejo  @jafarre-viewnext

### LINKED ISSUE #415: Support Jira Versions for document generation
In order to allow to split features into different versions, and to document only what the product has versus everything it is in Jira, we need to adapt the MRO document generation part to populate the documentation only with the correct data.

### PR DESCRIPTION
**Pull request by stitakis**
Closes #414
Closes #415

### CODE REVIEW

File: src/org/ods/orchestration/service/leva/ProjectDataBitbucketRepository.groovy
Comment (stitakis) [inline src/org/ods/orchestration/service/leva/ProjectDataBitbucketRepository.groovy:31]: not a big deal more a question of code style... there is no need to use this `this.`

File: src/org/ods/orchestration/util/DocumentHistory.groovy
Comment (stitakis) [inline src/org/ods/orchestration/util/DocumentHistory.groovy:20]: to improve encapsulation the following internal class members could be set to private

File: src/org/ods/orchestration/util/DocumentHistory.groovy
Comment (stitakis) [inline src/org/ods/orchestration/util/DocumentHistory.groovy:145]: to improve encapsulation all the following `protected method in this class could be set to `private`. Inheritance is not needed atm as there is no child class of `DocumentHistory`

File: src/org/ods/orchestration/util/DocumentHistoryEntry.groovy
Comment (stitakis) [inline src/org/ods/orchestration/util/DocumentHistoryEntry.groovy:11]: this fix encapsulation just to allow unit test to be able to change the... a solution be to convert it to 'private String rational' and additionally this method needs to be added to avoid breaking the unit test:
```
@NonCPS
    String setRational(rational) {
        return this.rational = rational
    }
```

**Review by stitakis — COMMENTED** [review]
I found some minor improvement that from code style point of view could be improved. Additionally encapsulation could be improved in 2 classes. Please see my comments. 
My IDE reports that unit test coverage of the orchestration package is around 70% (class 70%, method 72%, line 68%). I would said it acceptable. Some more unit test could be added before merging to master.

File: src/org/ods/orchestration/util/DocumentHistory.groovy
Comment (stitakis) [inline src/org/ods/orchestration/util/DocumentHistory.groovy:30]: to improve encapsulation the following internal class members could be set to private

**Comment by stitakis** [comment]
@metmajer I added my review comments... overall looks good to me. I would approve this change, but it looks like I cannot approve it because I created this PR.

File: src/org/ods/orchestration/util/DocumentHistory.groovy
Comment (martsec) [reply src/org/ods/orchestration/util/DocumentHistory.groovy:30]: As a comment, I have seen that groovy ignores private/protected tags. You can access anything from outside

**Comment by metmajer** [comment]
@stitakis I have requested a review from my side before we merge. Will do so asap.

File: src/org/ods/orchestration/InitStage.groovy
Comment (metmajer) [inline src/org/ods/orchestration/InitStage.groovy:327]: This must not happen here, but in Project.load()

File: src/org/ods/orchestration/service/leva/ProjectDataBitbucketRepository.groovy
Comment (metmajer) [reply src/org/ods/orchestration/service/leva/ProjectDataBitbucketRepository.groovy:31]: @stitakis we use `this` references in a larger portion of the Release Manager, which aids readability at least for me. `this` shows me that you access a member, vs. without `this` can only reference a method argument.

File: src/org/ods/orchestration/service/leva/ProjectDataBitbucketRepository.groovy
Comment (metmajer) [inline src/org/ods/orchestration/service/leva/ProjectDataBitbucketRepository.groovy:37]: Typo

File: src/org/ods/orchestration/usecase/JiraUseCase.groovy
Comment (metmajer) [inline src/org/ods/orchestration/usecase/JiraUseCase.groovy:120]: `this.project.versionFromReleaseStatusIssue` runs a several queries against Jira. Why not use the `Project.getVersionName()`?

File: src/org/ods/orchestration/util/Project.groovy
Comment (metmajer) [inline src/org/ods/orchestration/util/Project.groovy:430]: This line changes the expectation that only tests are processed which are DONE. Are we fine with this change?

**Review by metmajer — CHANGES_REQUESTED** [review]
@jorge-romero @jafarre-viewnext @hrcornejo @louis-sanchez please go over the feedback and let me know where help is needed.

File: src/org/ods/orchestration/util/DocumentHistoryEntry.groovy
Comment (metmajer) [reply src/org/ods/orchestration/util/DocumentHistoryEntry.groovy:11]: not sure if that's needed @stitakis... in the end, Groovy automatically creates getters and setters for public fields, which is the case here. Unless we need a specific setter or getter that is off-standard, no need to create a longer codebase. agreed?

File: src/org/ods/orchestration/usecase/JiraUseCase.groovy
Comment (metmajer) [reply src/org/ods/orchestration/usecase/JiraUseCase.groovy:120]: We should not use `this.project.versionFromReleaseStatusIssue` or other methods which load data from Jira more then once. Currently, is it in use when loading versioned data in Project.groovy.

Comment (s2oBCN) [reply src/org/ods/orchestration/usecase/JiraUseCase.groovy:120]: As this is Groovy, and we can not hide things with 'private', another solution that I saw is to create a sub-package with the name "private", for the things that we don't want other code to use.

File: src/org/ods/orchestration/util/DocumentHistory.groovy
Comment (s2oBCN) [reply src/org/ods/orchestration/util/DocumentHistory.groovy:30]: Groovy treats visibility keywords as a hint for a programmer. I think as a good practice to use this modifers for readability reasons. 
Maybe we have to define some groovy-coding-style guide... think about:
- https://dzone.com/articles/groovy-coding-style  
- https://google.github.io/styleguide/javaguide.html
- https://groovy-lang.org/style-guide.html

File: src/org/ods/orchestration/util/Project.groovy
Comment (s2oBCN) [reply src/org/ods/orchestration/util/Project.groovy:431]: This change was done by Marti on 14/08. Now we have all the manual tests passed. I think it should be ok.

Comment (metmajer) [reply src/org/ods/orchestration/util/Project.groovy:468]: @s2oBCN it changes the logic which was based on a constraint of mine. I think it's worth having a discussion.

File: src/org/ods/orchestration/usecase/JiraUseCase.groovy
Comment (metmajer) [reply src/org/ods/orchestration/usecase/JiraUseCase.groovy:120]: @s2oBCN @jorge-romero I'd consider it good practice to load any Jira data in `Project.load` and not load any data multiple times. Unfortunately, versions are loaded in Project.groovy and JiraUseCase.groovy. This needs to be consolidated in the next version. I'll add `FIXME` comments to remind us.

**Pull request merged**
